### PR TITLE
[1.16.1]Allow HoverEvent creation from ItemStack Entity

### PR DESCRIPTION
**Pull request by yannicklamprecht**
Allows getting a HoverEvent from ItemStack and Entity.

`Bukkit.hoverEventOf(itemStack)`
`Bukkit.hoverEventOf(entity)`

Test plugin: https://github.com/yannicklamprecht/PaperPRTestPlugins/blob/master/src/main/java/com/github/yannicklamprecht/paper/pr/test/testplugins/HoverTestCommand.java

### CODE REVIEW

**Comment by Machine-Maker** [comment]
Shouldn't this be added a part of the whole component system?

**Comment by Proximyst** [comment]
That would require us having to be able to patch more dependencies of Spigot for an admittedly small feature.
Using this way or the way I proposed (#3266) would not require that.

**Comment by Machine-Maker** [comment]
Will spigot not implement this as part of their component system if pr'd? Like, it'd just make more sense if it was there. 

If spigot won't implement it then yeah, either a Joiner or like this.

**Comment by yannicklamprecht** [comment]
> Will spigot not implement this as part of their component system if pr'd? Like, it'd just make more sense if it was there.
> 
> If spigot won't implement it then yeah, either a Joiner or like this.

I guess the problem with the component system is, that you cannot depend on Bukkits ItemStack. That's probably why it wasn't added there.

Will add Entity and Advancement as well if possible.

**Comment by Machine-Maker** [comment]
> I guess the problem with the component system is, that you cannot depend on Bukkits ItemStack. That's probably why it wasn't added there.

oh yeah, didn't think about that.

**Comment by yannicklamprecht** [comment]
Added methods for `Entity`

File: Spigot-Server-Patches/0524-Create-HoverEvent-from-ItemStack-Entity.patch [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     }
++
++    @Override
++    public HoverEvent hoverEventOf(ItemStack itemStack) {
++        String compoundString = CraftItemStack.asNMSCopy(itemStack).save(new NBTTagCompound()).toString();
++        return new HoverEvent(HoverEvent.Action.SHOW_ITEM, new BaseComponent[]{new TextComponent(compoundString)});
++    }
++
++    @Override
++    public HoverEvent hoverEventOf(Entity entity, String customName) {
```
Comment (yannicklamprecht) [inline Spigot-Server-Patches/0524-Create-HoverEvent-from-ItemStack-Entity.patch:41]: Better to pass a Base component instead? Or keep it legacy?

Comment (MiniDigger) [reply Spigot-Server-Patches/0524-Create-HoverEvent-from-ItemStack-Entity.patch:41]: why not both? :D

we should definitely be exposing components where components are accepted

Comment (yannicklamprecht) [reply Spigot-Server-Patches/0524-Create-HoverEvent-from-ItemStack-Entity.patch:41]: Done.

**Review by Proximyst — APPROVED** [review]
Other than the extra whitespace and lack thereof, this patch looks great!

**Comment by yannicklamprecht** [comment]
> Other than the extra whitespace and lack thereof, this patch looks great!

Hope I fixed the whitespaces. Yeah where to much.

**Comment by Phoenix616** [comment]
This should probably use bungee-chat's [ContentEntity and ContentItem](https://github.com/SpigotMC/BungeeCord/blob/master/chat/src/main/java/net/md_5/bungee/api/chat/HoverEvent.java#L202-L334) instead.

**Comment by yannicklamprecht** [comment]
> This should probably use bungee-chat's [ContentEntity and ContentItem](https://github.com/SpigotMC/BungeeCord/blob/master/chat/src/main/java/net/md_5/bungee/api/chat/HoverEvent.java#L202-L334) instead.

Was not present at that time. Will not change that now. In a month probably.

**Comment by aikar** [comment]
forgot to switch this to master before deleting the 1.16 branch, please reopen on master

**Comment by yannicklamprecht** [comment]
> This should probably use bungee-chat's [ContentEntity and ContentItem](https://github.com/SpigotMC/BungeeCord/blob/master/chat/src/main/java/net/md_5/bungee/api/chat/HoverEvent.java#L202-L334) instead.

Not usable due that fact that Enchantment id's are deprecated there.

**Comment by yannicklamprecht** [comment]
@MiniDigger here is your link. https://github.com/PaperMC/Paper/pull/3835 😅

**Pull request closed without merging**